### PR TITLE
Use single table inheritance to unify criterion types

### PR DESCRIPTION
**Pull request by david-yz-liu**
## Motivation and Context

The code for criteria is overly complex because each criterion type is stored in a separate table. E.g., we needed a `Assignment#get_criteria` method instead of a simple association, and to pass around both criterion id and type to disambiguate between different criterion types.

## Your Changes

**Description**: Unify the criteria tables using single table inheritance.

**Type of change** (select all that apply):

 

- [x] Refactoring (internal change to codebase, without changing functionality)

## Testing

Ensured that existing tests all pass, and manual testing in the UI for places where criteria are used.

## Questions and Comments (if applicable)

This is the work of @ashih2018 (I'm just merging).

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have updated the Changelog.md file.

### CODE REVIEW

File: app/controllers/results_controller.rb [hunk 1]
```diff
@@ -204,13 +206,13 @@ def show
         if assignment.assign_graders_to_criteria && current_user.ta?
           assigned_criteria = current_user.criterion_ta_associations
                                           .where(assessment_id: assignment.id)
-                                          .pluck(:criterion_type, :criterion_id)
-                                          .map { |t, id| "#{t}-#{id}" }
+                                          .pluck(:criterion_id)
+                                          .map(&:to_s)
```
Comment (david-yz-liu) [inline app/controllers/results_controller.rb:210]: I think I changed this line in the merge from master, but we probably don't need the `to_s` now that we're working with ids. Try making `assigned_criteria` an array of ints, and modify how it's used below.

File: app/controllers/results_controller.rb [hunk 2]
```diff
@@ -241,8 +243,11 @@ def show
         end
 
         # Totals
-        data[:assignment_max_mark] =
-          result.is_a_review? ? assignment.pr_assignment.max_mark(:peer) : marks_map.map { |h| h['max_mark'] }.sum
+        if result.is_a_review?
+          data[:assignment_max_mark] = assignment.pr_assignment.max_mark(:peer_visible)
+        else
+          data[:assignment_max_mark] = marks_map.map { |h| h['max_mark'] }.sum
```
Comment (david-yz-liu) [inline app/controllers/results_controller.rb:249]: Modify this to call `assignment.max_mark` (I don't know why this wasn't used before).

File: app/controllers/results_controller.rb [hunk 3]
```diff
@@ -542,8 +547,7 @@ def update_mark
     mark_value = params[:mark].blank? ? nil : params[:mark].to_f
 
     result_mark = result.marks.find_or_create_by(
-      markable_id: params[:markable_id],
-      markable_type: params[:markable_type]
+      criterion_id: params[:criterion_id]
```
Comment (david-yz-liu) [inline app/controllers/results_controller.rb:550]: Make this call one line now that it's shorter.

File: app/controllers/results_controller.rb [hunk 4]
```diff
@@ -688,10 +693,10 @@ def view_marks
     if @result.is_a_review?
       if @current_user.is_reviewer_for?(@assignment.pr_assignment, @result) ||
           !@grouping.membership_status(current_user).nil? || !@current_user.student?
-        @mark_criteria = @assignment.get_criteria(:peer)
+        @mark_criteria = @assignment.criteria.where(&:peer_visible)
```
Comment (david-yz-liu) [inline app/controllers/results_controller.rb:696]: Create an association `peer_criteria` (analogous to `ta_criteria`) and use it here.

File: app/helpers/automated_tests_helper.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 module AutomatedTestsHelper
   def extra_test_group_schema(assignment)
-    criterion_names, criterion_disambig = assignment.get_criteria(:ta).map do |c|
+    criterion_names, criterion_disambig = assignment.ta_criteria.map do |c|
       [c.name, "#{c.id}_#{c.class.name}"]
```
Comment (david-yz-liu) [inline app/helpers/automated_tests_helper.rb:4]: This data can be simplified, make the second element just the `c.id` (and not a string).

File: app/helpers/automated_tests_helper.rb [hunk 6]
```diff
@@ -48,12 +48,11 @@ def update_test_groups_from_specs(assignment, test_specs)
         display_output = extra_data_specs['display_output'] || TestGroup.display_outputs.keys.first
         test_group_name = extra_data_specs['name'] || TestGroup.model_name.human
         criterion_id = nil
-        criterion_type = nil
         if !extra_data_specs['criterion'].nil? && extra_data_specs['criterion'].include?('_')
```
Comment (david-yz-liu) [inline app/helpers/automated_tests_helper.rb:51]: Remove the second clause. Just set `criterion_id = extra_data_specs['criterion']` (this goes with my previous comment).

File: app/helpers/random_assign_helper.rb [hunk 7]
```diff
@@ -155,7 +155,7 @@ def swap(arr, i, j)
   def save_peer_reviews(pr_assignment)
     return if @reviewers.empty? || @reviewees.empty?
 
-    assignment_criteria = pr_assignment.get_criteria(:peer)
+    assignment_criteria = pr_assignment.criteria.where(:peer_visible => true)
```
Comment (david-yz-liu) [inline app/helpers/random_assign_helper.rb:158]: Simplify to `.peer_criteria` when you create the association.

File: app/models/assignment.rb [hunk 8]
```diff
@@ -268,10 +258,10 @@ def display_for_note
   end
 
   # Returns the maximum possible mark for a particular assignment
-  def max_mark(user_visibility = :ta)
+  def max_mark(user_visibility = :ta_visible)
     # TODO: sum method does not work with empty arrays. Consider updating/replacing gem:
     #       see: https://github.com/thirtysixthspan/descriptive_statistics/issues/44
-    max_marks = get_criteria(user_visibility).map(&:max_mark)
+    max_marks = criteria.where(user_visibility => true).map(&:max_mark)
```
Comment (david-yz-liu) [inline app/models/assignment.rb:264]: This method can be simplified now that it's a single query. Before, `get_criteria` returned an array, which why there is the `map` and the comment about summing an empty array.

But now you should be able to call `criteria.where(user_visibility => true).sum(:max_mark)` (or something similar) to grab the sum of the criteria max_marks, even when there are none.

File: app/models/assignment.rb [hunk 9]
```diff
@@ -525,7 +515,7 @@ def summary_json(user)
       if self.assign_graders_to_criteria
         assigned_criteria = user.criterion_ta_associations
                                 .where(assessment_id: self.id)
-                                .pluck(:criterion_type, :criterion_id)
+                                .pluck(:criterion_id)
                                 .map { |t, id| "#{t}-#{id}" }
```
Comment (david-yz-liu) [inline app/models/assignment.rb:519]: This is another line that is no longer necessary: `assigned_criteria` should just be an array of ids. (Modify usage as necessary)

File: app/models/assignment.rb [hunk 10]
```diff
@@ -625,7 +615,7 @@ def summary_csv(user)
     end
 
     headers = [['User name', 'Group', 'Final grade'], ['', 'Out of', self.max_mark]]
-    criteria = self.get_criteria(:ta)
+    criteria = self.ta_criteria
```
Comment (david-yz-liu) [inline app/models/assignment.rb:618]: If the only usage of this variable is on the previous line, remove this variable and just use `self.ta_criteria` on the line below.

File: app/models/assignment.rb [hunk 11]
```diff
@@ -669,51 +659,13 @@ def get_marks_list(submission)
   def next_criterion_position
     # We're using count here because this fires off a DB query, thus
     # grabbing the most up-to-date count of the criteria.
-    get_criteria.count > 0 ? get_criteria.last.position + 1 : 1
+    criteria.count > 0 ? criteria.last.position + 1 : 1
   end
 
   # Returns a filtered list of criteria.
   def get_criteria(user_visibility = :all, type = :all, options = {})
```
Comment (david-yz-liu) [inline app/models/assignment.rb:666]: You can delete this function altogether!

File: app/models/checkbox_criterion.rb [hunk 12]
```diff
@@ -57,7 +48,7 @@ def self.create_or_update_from_csv_row(row, assignment)
 
     # If a CheckboxCriterion with the same name exists, load it up. Otherwise,
     # create a new one.
-    criterion = assignment.get_criteria(:all, :checkbox).find_or_create_by(name: name)
+    criterion = assignment.criteria.where(type: 'CheckboxCriterion').find_or_create_by(name: name)
```
Comment (david-yz-liu) [inline app/models/checkbox_criterion.rb:51]: Ah, simplify this by removing the `where` and doing `find_or_create_by(name: ..., type: 'CheckboxCriterion')`

File: app/models/criterion.rb [hunk 13]
```diff
@@ -1,46 +1,65 @@
...
   validates_presence_of :name
   validates_uniqueness_of :name, scope: :assessment_id
 
   validates_presence_of :max_mark
   validates_numericality_of :max_mark, greater_than: 0
 
   has_many :criteria_assignment_files_joins,
-           as: :criterion,
            dependent: :destroy
+
```
Comment (david-yz-liu) [inline app/models/criterion.rb:27]: Delete this line

File: app/models/criterion.rb [hunk 14]
```diff
@@ -1,46 +1,65 @@
...
   has_many :assignment_files,
            through: :criteria_assignment_files_joins
   accepts_nested_attributes_for :criteria_assignment_files_joins, allow_destroy: true
 
   validate :results_unreleased?
   validate :visible?
 
-  self.abstract_class = true
+  has_many :levels, -> { order(:mark) }, inverse_of: :criterion, dependent: :destroy, autosave: true
```
Comment (david-yz-liu) [inline app/models/criterion.rb:35]: Hmm, this association still belongs in the `RubricCriterion` model, I think. I can't remember if you ran into an issue keeping it in `RubricCriterion`---if so, let me know, and if not, move it back there.

(Similar with line below)

File: app/models/criterion.rb [hunk 15]
```diff
@@ -1,46 +1,65 @@
...
+  accepts_nested_attributes_for :levels, allow_destroy: true
+
+  def update_assigned_groups_count
+    result = []
+    criterion_ta_associations.each do |cta|
```
Comment (david-yz-liu) [inline app/models/criterion.rb:40]: As with a previous comment, use `flat_map` to simplify.

File: app/models/criterion.rb [hunk 16]
```diff
@@ -85,23 +100,14 @@ def self.assign_tas(criterion_ids_types, ta_ids, assignment)
     # CriterionTaAssociation.create when the PG driver supports bulk create,
     # then remove the activerecord-import gem.
     CriterionTaAssociation.import(columns, values, validate: false)
-
```
Comment (david-yz-liu) [inline app/models/criterion.rb:88]: Revert this change

File: app/models/criterion.rb [hunk 17]
```diff
@@ -118,23 +124,20 @@ def self.update_assigned_groups_counts(assignment)
...
-    [RubricCriterion, FlexibleCriterion, CheckboxCriterion].each do |klass|
-      records = klass.where(assessment_id: assignment.id)
-                     .pluck_to_hash
-                     .map do |h|
-        { **h.symbolize_keys, assigned_groups_count: counts[[h['id'], klass.to_s]] || 0 }
-      end
-      unless records.empty?
-        klass.upsert_all(records)
-      end
+    records = Criterion.where(assessment_id: assignment.id)
```
Comment (david-yz-liu) [inline app/models/criterion.rb:133]: Simplify this: `assignment.criteria`

File: app/models/flexible_criterion.rb [hunk 18]
```diff
@@ -61,7 +47,7 @@ def self.create_or_update_from_csv_row(row, assignment)
     name = working_row.shift
     # If a FlexibleCriterion with the same name exits, load it up.  Otherwise,
     # create a new one.
-    criterion = assignment.get_criteria(:all, :flexible).find_or_create_by(name: name)
+    criterion = assignment.criteria.where(type: 'FlexibleCriterion').find_or_create_by(name: name)
```
Comment (david-yz-liu) [inline app/models/flexible_criterion.rb:50]: See my comment in the `CheckboxCriterion` model.

File: app/models/result.rb [hunk 19]
```diff
@@ -80,13 +80,13 @@ def get_subtotal(assignment: nil)
...
-        user_visibility = :peer
+        user_visibility = :peer_visible
         assignment = assignment.pr_assignment
       else
-        user_visibility = :ta
+        user_visibility = :ta_visible
       end
-      criteria = assignment.get_criteria(user_visibility).map { |c| [c.class.to_s, c.id] }
-      marks_array = (marks.to_a.select { |m| criteria.member? [m.markable_type, m.markable_id] }).map &:mark
+      criteria = assignment.criteria.where(user_visibility => true).map(&:id)
```
Comment (david-yz-liu) [inline app/models/result.rb:88]: Don't use `map(&:id)`, use `.ids`. Also rename the variable to `criterion_ids`.

File: app/models/result.rb [hunk 20]
```diff
@@ -204,7 +204,7 @@ def create_marks
   def mark_hash
     Hash[
       marks.map do |mark|
-        ["#{mark.markable_type}-#{mark.markable_id}",
+        ["#{mark.criterion.type}-#{mark.criterion_id}",
```
Comment (david-yz-liu) [inline app/models/result.rb:207]: This is another place where you should be able to remove the `criterion.type` part (check usages!).

File: app/models/result.rb [hunk 21]
```diff
@@ -221,21 +221,21 @@ def check_for_released
     true
   end
 
-  def check_for_nil_marks(user_visibility = :ta)
+  def check_for_nil_marks(user_visibility = :ta_visible)
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/models/result.rb:224]: (ignore)

File: app/models/result.rb [hunk 22]
```diff
@@ -221,21 +221,21 @@ def check_for_released
...
 
     # peer review result is a special case because when saving a pr result
     # we can't pass in a parameter to the before_save filter, so we need
     # to manually determine the visibility. If it's a pr result, we know we
     # want the peer-visible criteria
-    visibility = is_a_review? ? :peer : user_visibility
+    visibility = is_a_review? ? :peer_visible : user_visibility
 
-    criteria = submission.assignment.get_criteria(visibility).map { |c| [c.class.to_s, c.id] }
+    criteria = submission.assignment.criteria.where(visibility => true).map(&:id)
```
Comment (david-yz-liu) [inline app/models/result.rb:234]: Use `.ids`

File: app/models/rubric_criterion.rb [hunk 23]
```diff
@@ -98,7 +69,7 @@ def self.create_or_update_from_csv_row(row, assignment)
     working_row = row.clone
     name = working_row.shift
 
-    criterion = assignment.get_criteria(:all, :rubric).find_or_create_by(name: name)
+    criterion = assignment.criteria.where(type: 'RubricCriterion').find_or_create_by(name: name)
```
Comment (david-yz-liu) [inline app/models/rubric_criterion.rb:72]: See comment about `CheckboxCriterion`.

File: app/models/submission.rb [hunk 24]
```diff
@@ -145,12 +145,13 @@ def set_autotest_marks
...
-        final_mark = mark.markable.max_mark
+        final_mark = mark.criterion.max_mark
       else
-        final_mark = (all_marks_earned / all_marks_total * mark.markable.max_mark).round(2)
-        if mark.markable.instance_of? RubricCriterion
+        final_mark = (all_marks_earned / all_marks_total * mark.criterion.max_mark).round(2)
+        if mark.criterion.instance_of? RubricCriterion
           # find the nearest mark associated to a level
-          nearest_mark = (final_mark / mark.markable.weight.to_f).round * mark.markable.weight
+          # this logic is incorrect due to having an arbitrary number of levels in a rubric criterion
```
Comment (david-yz-liu) [inline app/models/submission.rb:153]: Nice catch. Mark this as a TODO comment, and also create a new issue to mark this problem. (The problem predates this PR.)

File: app/models/ta.rb [hunk 25]
```diff
@@ -36,28 +36,25 @@ def percentage_grades_array(assignment)
     grades = []
 
     if assignment.assign_graders_to_criteria
-      criteria_data = self.criterion_ta_associations.where(assessment_id: assignment.id)
-                          .pluck(:criterion_type, :criterion_id)
-      out_of = criteria_data.sum do |criterion_type, criterion_id|
-        criterion_type.constantize.find(criterion_id).max_mark
+      criteria_data = self.criterion_ta_associations.where(assessment_id: assignment.id).pluck(:criterion_id)
```
Comment (david-yz-liu) [inline app/models/ta.rb:39]: You can simplify this logic even further. Something like `self.criteria.where('criteria.assessment_id': assignment.id).sum(:max_mark)`.

File: lib/tasks/groups.rake [hunk 26]
```diff
@@ -4,9 +4,9 @@ namespace :db do
   task :groups => :environment do
     puts 'Assign Groups/Students for Assignments'
     Faker::Config.random = Random.new(42) # seeds the random number generator so Faker output is deterministic
-    students = Student.all
+    students = Student.limit(50)
```
Comment (david-yz-liu) [inline lib/tasks/groups.rake:7]: Revert the changes in this file.

File: app/helpers/automated_tests_helper.rb (hunk 6 above)
Comment (ashih2018) [reply app/helpers/automated_tests_helper.rb:51]: What do you mean by this goes with your previous comment?

File: app/models/criterion.rb (hunk 14 above)
Comment (ashih2018) [reply app/models/criterion.rb:34]: I'll keep you updated when I start doing UI testing!

File: app/models/ta.rb (hunk 25 above)
Comment (ashih2018) [reply app/models/ta.rb:39]: We can't change this logic because we use `criteria_data` later when finding mark data, so we still want to keep track of the ids. Rather we should probably just change the name to `criteria_ids`.

File: spec/models/assignment_spec.rb [hunk 27]
```diff
@@ -12,12 +12,6 @@
     it { is_expected.to have_many(:section_due_dates) }
     it { is_expected.to accept_nested_attributes_for(:section_due_dates) }
     it { is_expected.to have_one(:assignment_stat).dependent(:destroy) }
-    it do
-      is_expected.to have_many(:rubric_criteria).dependent(:destroy).order(:position)
```
Comment (david-yz-liu) [inline spec/models/assignment_spec.rb:16]: You should add in a new test for `is_expected.to have_many(:criteria)` with the same options, as well as `ta_criteria` and `peer_criteria` associations.

File: lib/tasks/groups.rake [hunk 28]
```diff
@@ -6,7 +6,7 @@ namespace :db do
     Faker::Config.random = Random.new(42) # seeds the random number generator so Faker output is deterministic
     students = Student.all
     Assignment.all.each do |assignment|
-      num_groups = (assignment.short_identifier == 'A1' || assignment.short_identifier == 'A3') ? students.length : 15
+      num_groups = assignment.short_identifier == 'A1' || assignment.short_identifier == 'A3' ? students.length : 5
```
Comment (david-yz-liu) [inline lib/tasks/groups.rake:9]: Revert the changes in this file: 5 -> 15, but also how you fixed the Hound comment. Basically, this file shouldn't show any changes in the diff with master.

File: spec/controllers/criteria_controller_spec.rb [hunk 29]
```diff
@@ -858,78 +859,56 @@
...
 
-        expect(assignment.get_criteria.map(&:name)).not_to include('cr70')
+        expect(assignment.criteria.map(&:name)).not_to include('cr70')
       end
 
       it 'does not create criteria that have unmatched keys / more keys than required' do
+        expect(assignment.criteria.where(type: 'RubricCriterion').length).to eq(1)
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: partially_valid_file }
-
-        criteria = assignment.get_criteria(:all, :rubric).first
```
[the remark this answers is not in the document]
Comment (ashih2018) [reply spec/controllers/criteria_controller_spec.rb:909]: It was because there are errors that are raised so `ActiveRecord::Rollback` always gets called so the new criteria is never created. Because of this we only get returned the criteria created in the `before each` call at the beginning of this context, so I edited the test to check that there is an error raised for a 'partially invalid file'.

[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/controllers/criteria_controller_spec.rb:909]: Okay that makes sense, thanks!
